### PR TITLE
Implement build oni the provider side.

### PR DESCRIPTION
**Pull request by marmistrz**
Known limitations:
* the Makefile support is rudimentary (see the tutorial)
* cmake is pretty sensitive to the binary locations (symlinks and wrappers don't work very well with cmake), so we need to use a portable binary distribution for the gu image. With docker sessions we'll be able to call arbitrary commands from the docker image, so this will be no longer needed.
* we only support `.tar` files because gu supports only tars.
* we deploy sequentially on every provider. We could 
    * just create threads/use rayon
    * make the code more even more async-oriented [difficult] 
    * use the wait command of gu [available only for the docker environment, currently doesn't work with the host-direct environment]

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -36,7 +36,11 @@ fn run() -> Fallible<()> {
     let config = JobConfig::from_file(&opt.jobconfig).context("reading job config")?;
 
     let mgr = SessionMPI::init(opt.hub)?;
-    println!("HOSTFILE:\n{}", mgr.hostfile()?);
+
+    if let Some(sources) = config.sources {
+        mgr.deploy(opt.jobconfig.parent().unwrap(), &sources)
```
Comment (kubkon) [inline src/main.rs]: `unwrap` here seems dangerous. What if we run from a container such as Docker and have everything under `/`? Then `parent()` would return `None`.

Comment (marmistrz) [reply src/main.rs]: Should `opt.jobconfig.parent()` be `None`, it wouldn't be a valid file path, so we'd already have gotten an error in main.rs:36. 

Should I add a comment about it?

Comment (kubkon) [reply src/main.rs]: Yeah, that'd be very helpful, thanks!

Comment (marmistrz) [reply src/main.rs]: Added.

Comment (kubkon) [reply src/main.rs]: :+1:

**Pull request merged**